near-contract-standards/src/fungible_token/implementation.rs

```diff
@@ -0,0 +1,256 @@
+            total_supply: 0,
+            account_storage_usage: 0,
+        };
+        let initial_storage_usage = env::storage_usage();
+        let tmp_account_id = unsafe { String::from_utf8_unchecked(vec![b'a'; 64]) };
+        this.accounts.insert(&tmp_account_id, &0u128);
+        this.account_storage_usage = env::storage_usage() - initial_storage_usage;
+        this.accounts.remove(&tmp_account_id);
+        this
+    }
+
+    pub fn internal_deposit(&mut self, account_id: &AccountId, amount: Balance) {
+        let balance = self.accounts.get(&account_id).expect("The account is not registered");
+        if let Some(new_balance) = balance.checked_add(amount) {
+            self.accounts.insert(&account_id, &new_balance);
+            self.total_supply =
+                self.total_supply.checked_add(amount).expect("Total supply overflow");
+        } else {
+            env::panic(b"Balance overflow");
+        }
+    }
+
+    pub fn internal_withdraw(&mut self, account_id: &AccountId, amount: Balance) {
+        let balance = self.accounts.get(&account_id).expect("The account is not registered");
+        if let Some(new_balance) = balance.checked_sub(amount) {
+            self.accounts.insert(&account_id, &new_balance);
+            self.total_supply =
+                self.total_supply.checked_sub(amount).expect("Total supply overflow");
+        } else {
+            env::panic(b"The account doesn't have enough balance");
+        }
+    }
+
+    pub fn internal_transfer(
+        &mut self,
+        sender_id: &AccountId,
+        receiver_id: &AccountId,
+        amount: Balance,
+        memo: Option<String>,
+    ) {
+        assert_ne!(sender_id, receiver_id, "Sender and receiver should be different");
+        assert!(amount > 0, "The amount should be a positive number");
+        self.internal_withdraw(sender_id, amount);
+        self.internal_deposit(receiver_id, amount);
+        log!("Transfer {} from {} to {}", amount, sender_id, receiver_id);
+        if let Some(memo) = memo {
+            log!("Memo: {}", memo);
+        }
+    }
+
+    pub fn internal_register_account(&mut self, account_id: &AccountId) {
+        if self.accounts.insert(&account_id, &0).is_some() {
+            env::panic(b"The account is already registered");
+        }
+    }
+}
+
+impl FungibleTokenCore for FungibleToken {
+    fn ft_transfer(&mut self, receiver_id: ValidAccountId, amount: U128, memo: Option<String>) {
+        assert_one_yocto();
+        let sender_id = env::predecessor_account_id();
+        let amount = amount.into();
+        self.internal_transfer(&sender_id, receiver_id.as_ref(), amount, memo);
+    }
+
+    fn ft_transfer_call(
+        &mut self,
+        receiver_id: ValidAccountId,
+        amount: U128,
+        memo: Option<String>,
+        msg: String,
+    ) -> Promise {
+        assert_one_yocto();
+        let sender_id = env::predecessor_account_id();
+        let amount = amount.into();
```

Comment:
A non-rust dev would find this line very strange (or even rust devs).  Though I love that the type inference from how amount used later works, it's still a little confusing.
```suggestion
```

Reply:
Perhaps it makes more sense to remove this line, then `amount.into()` in the function call.  This also removes two `into`s later.

Reply:
I've updated it to be `let amount: Balance = amount.into()` to give it a type.